cloudflare/resource_cloudflare_waf_group.go

```diff
@@ -0,0 +1,203 @@
+package cloudflare
+
+import (
+	"fmt"
+	"strings"
+
+	cloudflare "github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceCloudflareWAFGroup() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareWAFGroupCreate,
+		Read:   resourceCloudflareWAFGroupRead,
+		Update: resourceCloudflareWAFGroupUpdate,
+		Delete: resourceCloudflareWAFGroupDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareWAFGroupImport,
+		},
+
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"group_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"package_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"mode": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "on",
+				ValidateFunc: validation.StringInSlice([]string{"on", "off"}, false),
+			},
+		},
+	}
+}
+
+func resourceCloudflareWAFGroupRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+
+	groupID := d.Get("group_id").(string)
+	zoneID := d.Get("zone_id").(string)
+	packageID := d.Get("package_id").(string)
+
+	group, err := client.WAFGroup(zoneID, packageID, groupID)
+	if err != nil {
+		return (err)
+	}
+
+	// Only need to set mode as that is the only attribute that could have changed
+	d.Set("mode", group.Mode)
+	d.SetId(group.ID)
+
+	return nil
+}
+
+func resourceCloudflareWAFGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	groupID := d.Get("group_id").(string)
+	zoneID := d.Get("zone_id").(string)
+	packageID := d.Get("package_id").(string)
+	mode := d.Get("mode").(string)
+
+	// If no package ID is given try to resolve it
+	var pkgList []cloudflare.WAFPackage
+	if packageID == "" {
+		var err error
+		pkgList, err = client.ListWAFPackages(zoneID)
+		if err != nil {
+			return err
+		}
+	} else {
+		pkgList = append(pkgList, cloudflare.WAFPackage{ID: packageID})
+	}
+
+	for _, pkg := range pkgList {
+		var err error
+		var group cloudflare.WAFGroup
+
+		group, err = client.WAFGroup(zoneID, pkg.ID, groupID)
+
+		if err == nil {
```

Comment:
Instead of nesting here, let's flip the check and return early if there is an error. Make the flow a little more readable.

```go
if err != nil {
  // handle error
}

d.Set("group_id", group.ID)

// ...
```

Reply:
Updated